### ISSUE TITLE
Performance Question

### ISSUE THREAD

**Issue by owengalenjones** [everit-org/json-schema#469]
Hello, I have a large number of schemas linked by `$ref` 's with a combined disk size 1380KB and an object is 675B. In benchmarks it consistently takes around 1.4 seconds to validate the object. Is this an expected time? We were hoping to validate objects at runtime during client requests but this would make that impossible.

Are there any steps I can take to attempt to increase performance here?

Thanks!

**Comment by erosb** [comment]
Hello @owengalenjones , I can't advise on that based on this information. You can 
 * check if your schema has any large-scale `"oneOf"` applicator and if yes, if it is substitutable by an `"anyOf"` without loosing any  guarantees. (this can work because `"anyOf"` is short-circuit evaluated, unlike `"oneOf"`)
 * check with a profiler where is the most time spent 
 * compare the performance with other validator implementations

**Comment by owengalenjones** [comment]
@erosb thanks for the reply!

1. The schema I'm testing with is not failing so I would imagine short-circuiting wouldn't gain me anything?
2. I will check
3. So far this is the quickest of the validators I have examined, [networknt/json-schema-validator](https://github.com/networknt/json-schema-validator) took 9 seconds for the same validation 😨 but I guess I can try some of the others.

**Comment by owengalenjones** [comment]
As far as profiling, I'm not sure if it shows me anything:

<img width="1190" alt="Screen Shot 2022-08-31 at 1 59 20 PM" src="https://user-images.githubusercontent.com/690817/187781097-f26ec38f-bceb-4eb6-901a-0885497c1967.png">

I assume this is just standard recursive call stacks?

I ran the test 3 times to get data for profiling.

Hotspots:

<img width="1247" alt="Screen Shot 2022-08-31 at 2 01 11 PM" src="https://user-images.githubusercontent.com/690817/187781286-f9885920-a411-4b1c-acaa-00ee32e75dfd.png">

**Comment by erosb** [comment]
Yes, there can be many `ValidationException`s thrown, if there are many subschema mismatches during validation. But I can't see  `InternalValidationException` on the screenshots. Is it because its self-time is low, or do you use an old version of the library?

**Comment by owengalenjones** [comment]
Surprisingly with the case of the exceptions, I see no validation error messages when validation is complete. I am running tests of the schema against a sample JSON object that is valid / conforming to the schema.

We are using:

```
    <dependency>
      <groupId>com.github.erosb</groupId>
      <artifactId>everit-json-schema</artifactId>
      <version>1.9.2</version>
    </dependency>
```

**Comment by owengalenjones** [comment]
Ah I see that's slightly out of date. With 1.14.1 the same run is down to ~700ms.

**Comment by erosb** [comment]
Good. You may profile again, maybe there is a little chance to improve it.
I will get back to this issue tomorrow.

**Comment by owengalenjones** [comment]
Thanks for your assistance @erosb, despite the increase I'm still wondering if there is anything that my profile shows as a potential site of further improvment?

<img width="1260" alt="Screen Shot 2022-09-01 at 11 19 24 AM" src="https://user-images.githubusercontent.com/690817/187986403-34ee14c7-6bf4-4699-94f1-a406f023c7e5.png">

I'm not sure what would be expected or if there is anything that leaps out to you.

In testing the same collection of schemas and the JSON object with the node library Ajv, it only takes 40 ms.

**Comment by erosb** [comment]
Hello, my last guess: was the JVM warmed up? How many iterations of validation did you run?
